// @grammyjs/runner/out/source.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSource = void 0;
/**
 * Creates an update source based on the given update supplier.
 *
 * @param supplier An update supplier to use for requesting updates
 * @returns An update source
 */
function createSource(supplier) {
    let active = false;
    let controller;
    const listener = () => {
        active = false;
    };
    let w = worker();
    let pace = Infinity;
    async function* worker() {
        active = true;
        do {
            controller = new node_shim_js_1.AbortController();
            controller.signal.addEventListener("abort", listener);
            try {
                yield await supplier.supply(pace, controller.signal);
            }
            catch (e) {
                if (!controller.signal.aborted)
                    throw e;
                close();
                break;
            }
        } while (active);
    }
    function close() {
        active = false;
        controller.abort();
        w = worker();
        pace = Infinity;
    }
    return {
        generator: () => w,
        setGeneratorPace: (newPace) => (pace = newPace),
        isActive: () => active,
        close: () => close(),
    };
}
exports.createSource = createSource;
const node_shim_js_1 = require("./node-shim.js");
